keras_nlp/layers/bert_packer.py

```diff
@@ -0,0 +1,230 @@
+                - `"waterfall"`: The allocation of the budget is done using a
+                    "waterfall" algorithm that allocates quota in a
+                    left-to-right manner and fills up the buckets until we run
+                    out of budget. It support arbitrary number of segments.
+
+    Examples:
+
+    *Pack a single input for classification.*
+    >>> seq1 = [1, 2, 3, 4]  # Replace with a real input.
+    >>> packer = keras_nlp.layers.BertPacker(8, start_value=101, end_value=102)
+    >>> packer(seq1)
+    {
+        "tokens": <tf.Tensor: shape=(8,), dtype=int32,
+            numpy=array([101,   1,   2,   3,   4, 102,   0,   0], dtype=int32)>,
+        "padding_mask": <tf.Tensor: shape=(8,), dtype=float32,
+            numpy=array([1., 1., 1., 1., 1., 1., 0., 0.], dtype=float32)>,
+        "segment_ids": <tf.Tensor: shape=(8,), dtype=float32,
+            numpy=array([0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32)>,
+    }
+
+    *Pack multiple inputs for classification.*
+    >>> seq1 = [1, 2, 3, 4]  # Replace with a real input.
+    >>> seq2 = [11, 12, 13, 14]  # Replace with a real input.
+    >>> seq3 = [21, 12, 23, 24]  # Replace with a real input.
+    >>> packer = keras_nlp.layers.BertPacker(8, start_value=101, end_value=102)
+    >>> packer((seq1, seq2, seq3))
+    {
+        "tokens": <tf.Tensor: shape=(8,), dtype=int32,
+            numpy=array([101,   1,   2, 102,  11, 102,  21,  102], dtype=int32)>,
+        "padding_mask": <tf.Tensor: shape=(8,), dtype=float32,
+            numpy=array([1., 1., 1., 1., 1., 1., 1., 1.], dtype=float32)>,
+        "segment_ids": <tf.Tensor: shape=(8,), dtype=float32,
+            numpy=array([0., 0., 0., 0., 1., 1., 2., 2.], dtype=float32)>,
+    }
+
+    Reference:
+        [Devlin et al., 2018](https://arxiv.org/abs/1810.04805).
+    """
+
+    def __init__(
+        self,
+        sequence_length,
+        start_value,
+        end_value,
+        pad_value=None,
+        truncator="round_robin",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.sequence_length = sequence_length
+        if truncator not in ("round_robin", "waterfall"):
+            raise ValueError(
+                "Only 'round_robin' and 'waterfall' algorithms are "
+                "supported. Received %s" % truncator
+            )
+        self.truncator = truncator
+        self.start_value = start_value
+        self.end_value = end_value
+        self.pad_value = pad_value
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "sequence_length": self.sequence_length,
+                "start_value": self.start_value,
+                "end_value": self.end_value,
+                "pad_value": self.pad_value,
+                "truncator": self.truncator,
+            }
+        )
+        return config
+
+    def _sanitize_inputs(self, inputs):
+        """Force inputs to a list of rank 2 ragged tensors."""
```

Comment:
This currently doesn't support dense tensors beyond rank 1, as the whole point of this layer is really to be something that converts your ragged sequences to dense. This whole layer is just like a very advanced `ragged.to_tensor()` call.

I'm not sure we should support dense inputs here, as it would encourage an anti-pattern in usage. We don't want users to convert to dense before this call.

Also it's worth noting that this is supposed to be a "power user" layer. The main usage will be a BertTokenizer layer that combines WordPieceTokenizer and BertPacker into a single offering.

Reply:
Got it, then do we want to add an if-check to throw error if inputs are dense tensor with rank > 1? The current check will let this case slip through, and maybe error out at a certain line?

Reply:
sounds good, will add